File: 20221231/lru-cache-java/app/src/main/java/lru/cache/java/LRUCache.java
package lru.cache.java;

import java.util.HashMap;
import java.util.Map;

public class LRUCache {
    private final int capacity;
    private final Map<Integer, Node> nodes;
    private final MyLinkedList nodesWithPriority;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        nodes = new HashMap<>();
        nodesWithPriority = new MyLinkedList();
    }

    public int get(int key) {
        if (!nodes.containsKey(key)) {
            return -1;
        }

        Node node = nodes.get(key);
        updatePriority(node);
        return node.value();
    }

    private void updatePriority(Node node) {
        nodesWithPriority.remove(node);
        nodesWithPriority.addFirst(node);
    }

    public void put(int key, int value) {
        if (nodes.containsKey(key)) {
            Node node = nodes.get(key);
            update(node, value);
            return;
        }

        if (isFull()) {
            evictLeastUsedNode();
        }
        Node newest = new Node(key, value);
        nodesWithPriority.addFirst(newest);
        nodes.put(key, newest);
    }

    private void update(Node node, int value) {
        node.update(value);
        updatePriority(node);
    }

    private void evictLeastUsedNode() {
        Node leastUsed = nodesWithPriority.removeLast();
        nodes.remove(leastUsed.key());
    }

    public boolean isFull() {
        return capacity == nodes.size();
    }
}

File: 20221229/JavaScript/LRUCache.js
import LinkedList from './LinkedList';
import Node from './Node';

export default class LRUCache {
  static NOT_EXIST = -1;

  constructor(capacity) {
    this.capacity = capacity;
    this.map = new Map();
    this.list = new LinkedList();
  }

  put(key, value) {
    if (this.containsNode(key)) {
      this.update(key, value);
      return;
    }

    if (this.isFull()) {
      this.evictLeastUsed();
    }
    this.addNode({ key, value });
  }

  containsNode(key) {
    return this.map.has(key);
  }

  update(key, value) {
    const node = this.getNode(key);
    node.update({ value });

    this.updatePriority(node);
  }

  getNode(key) {
    return this.map.get(key);
  }

  updatePriority(node) {
    this.list.remove(node);
    this.list.addFirst(node);
  }

  isFull() {
    return this.capacity === this.map.size;
  }

  evictLeastUsed() {
    const leastUsed = this.list.removeLast();
    this.map.delete(leastUsed.key);
  }

  addNode({ key, value }) {
    const newest = new Node({ key, value });
    this.list.addFirst(newest);
    this.map.set(key, newest);
  }

  get(key) {
    if (!this.containsNode(key)) {
      return LRUCache.NOT_EXIST;
    }

    const node = this.getNode(key);
    this.updatePriority(node);
    return node.value;
  }
}
